### ISSUE TITLE
How to correctly pass an IActiveScriptSite parameter in Rust using windows-rs

### ISSUE THREAD

**Issue by joaoviictorti** [microsoft/windows-rs#3401]
### Suggestion

Hi, windows-rs team! 👋

I'm struggling to properly implement and pass an IActiveScriptSite parameter in Rust to use it with the `IActiveScript::SetScriptSite` method. While I understand how this is typically done in C and C++, I'm not sure how to translate this correctly into Rust using windows-rs.

## Context

In C, the process is straightforward: we create a struct with the required methods, associate a static vtable, and pass a pointer to the function. In C++, this is abstracted with classes. However, in Rust, the mechanism for implementing a COM interface like IActiveScriptSite is unclear to me.

The SetScriptSite method in windows-rs is defined as follows:
```rs
impl IActiveScript {
    pub unsafe fn SetScriptSite<P0>(&self, pass: P0) -> windows_core::Result<()>
    where
        P0: windows_core::Param<IActiveScriptSite>,
    {
        (windows_core::Interface::vtable(self).SetScriptSite)(
            windows_core::Interface::as_raw(self), 
            pass.param().abi(),
        )
        .ok()
    }
}
```

However, when trying to pass an implementation of IActiveScriptSite, I encounter issues like:

- `STATUS_ACCESS_VIOLATION` (exit code 0xc0000005): Likely due to an improperly configured vtable or associated methods.
- Type errors when attempting to implement `Param<IActiveScriptSite>`: It seems I’m either using the wrong type or initializing the structure incorrectly.

What is the correct way to implement and pass a COM interface like IActiveScriptSite in Rust?

OBS: The label is as a enhancement, I didn't see any other option, so I left it that way, sorry

**Comment by joaoviictorti** [comment]
And from what I’ve observed, the `IActiveScriptSite` has the `IActiveScriptSite_impl` structure, which is marked as pub but doesn’t seem to be exported for direct use in the library. This prevents me from implementing the interface correctly.

**Comment by joaoviictorti** [comment]
If I'm not mistaken, this would be the correct way to pass a parameter to `SetScripSite`? 

[How do I implement an existing COM interface?](https://kennykerr.ca/rust-getting-started/how-to-implement-com-interface.html#how-do-i-implement-an-existing-com-interface)

**Comment by joaoviictorti** [comment]
I managed to solve it, the problem was that the feature implement was missing, thanks anyway 🫡

**Comment by riverar** [comment]
I'm a bit confused because the errors you reported are runtime errors. Not having the `implement` feature would result in compile time errors.

At any rate, here's a working example. Just a heads up though: Windows Scripting Host and VBScript are actively being removed from upcoming versions of Windows.

```rust
use windows::{
    core::*,
    Win32::{
        Foundation::{BOOL, HWND},
        System::{
            Com::{
                CLSIDFromProgID, CoCreateInstance, CoInitialize, ITypeInfo, CLSCTX_INPROC_SERVER,
                EXCEPINFO,
            },
            Diagnostics::Debug::ActiveScript::{
                IActiveScript, IActiveScriptError, IActiveScriptParse64, IActiveScriptSite,
                IActiveScriptSiteWindow, IActiveScriptSiteWindow_Impl, IActiveScriptSite_Impl,
                SCRIPTSTATE, SCRIPTSTATE_CONNECTED, SCRIPTTEXT_ISVISIBLE,
            },
        },
    },
};
use {w, Interface, BSTR, PCWSTR};

#[implement(IActiveScriptSite, IActiveScriptSiteWindow)]
struct Site;

impl IActiveScriptSiteWindow_Impl for Site_Impl {
    fn GetWindow(&self) -> Result<HWND> {
        Ok(HWND::default())
    }

    fn EnableModeless(&self, _: BOOL) -> Result<()> {
        Ok(())
    }
}

impl IActiveScriptSite_Impl for Site_Impl {
    fn GetLCID(&self) -> Result<u32> {
        Ok(1033)
    }

    fn GetItemInfo(
        &self,
        _: &PCWSTR,
        _: u32,
        ppiunkitem: *mut Option<IUnknown>,
        ppti: *mut Option<ITypeInfo>,
    ) -> Result<()> {
        unsafe {
            if !ppiunkitem.is_null() {
                *ppiunkitem = None;
            }

            if !ppti.is_null() {
                *ppti = None;
            }
        }

        Ok(())
    }

    fn GetDocVersionString(&self) -> Result<BSTR> {
        Ok(BSTR::from("1.0"))
    }

    fn OnScriptTerminate(&self, _: *const VARIANT, _: *const EXCEPINFO) -> Result<()> {
        Ok(())
    }

    fn OnStateChange(&self, _: SCRIPTSTATE) -> Result<()> {
        Ok(())
    }

    fn OnScriptError(&self, _: Option<&IActiveScriptError>) -> Result<()> {
        Ok(())
    }

    fn OnEnterScript(&self) -> Result<()> {
        Ok(())
    }

    fn OnLeaveScript(&self) -> Result<()> {
        Ok(())
    }
}

fn main() -> Result<()> {
    unsafe {
        CoInitialize(None).ok()?;

        let engine: IActiveScript = CoCreateInstance(
            &CLSIDFromProgID(w!("VBScript"))?,
            None,
            CLSCTX_INPROC_SERVER,
        )?;

        let site: IActiveScriptSite = Site.into();
        engine.SetScriptSite(&site)?;

        let parser = engine.cast::<IActiveScriptParse64>()?;
        parser.InitNew()?;

        parser.ParseScriptText(
            w!(r#"MsgBox "Hello, Rust World!""#),
            None,
            None,
            None,
            0,
            0,
            SCRIPTTEXT_ISVISIBLE,
            std::ptr::null_mut(),
            std::ptr::null_mut(),
        )?;

        engine.SetScriptState(SCRIPTSTATE_CONNECTED)?;

        Ok(())
    }
}
```

**Comment by joaoviictorti** [comment]
Great, thanks! I had created my own implementation but was running into execution issues, which were the errors I mentioned earlier. I tried using #[implement()], but it wasn’t working because a specific feature was missing, and I couldn’t figure it out from the code. After checking the site kennykerr.ca/rust-getting-started/, I realized I needed to enable a specific feature for implement. Once I activated it, I was able to use #[implement()] and fix the rest of the code! 🫡
